cdc/generation.cc

```diff
@@ -374,4 +376,130 @@ void update_streams_description(
+        return std::none_of(times_and_ttls.begin(), times_and_ttls.end(),
+                [&] (std::pair<db_clock::time_point, int> p) {
+            // In this CDC log table there are no entries older than the table's creation time
+            // or (now - the table's ttl). We subtract 10s to account for some possible clock drift.
+            // If ttl is set to 0 then entries in this table never expire. In that case we look
+            // only at the table's creation time.
+            auto no_entries_older_than =
+                (p.second == 0 ? p.first : std::max(p.first, now - std::chrono::seconds(p.second)))
+                    - std::chrono::seconds(10);
+            return no_entries_older_than < ts;
+        });
+    });
+
+    // Find first generation timestamp such that some CDC log table may contain data in this generation.
+    // This and all later generations need to be written to the new streams table.
+    if (first != tss.begin()) {
+        --first;
+    }
+
+    cdc_log.info("First generation to rewrite: {}", *first);
+
+    co_await max_concurrent_for_each(first, tss.end(), 10, [&] (db_clock::time_point ts) -> future<> {
+        while (true) {
+            try {
+                co_return co_await do_update_streams_description(ts, *sys_dist_ks, { get_num_token_owners() });
+            } catch (...) {
+                cdc_log.warn("Failed to rewrite streams for generation {}: {}. Retrying in 60s.", ts, std::current_exception());
+            }
+            co_await sleep_abortable(std::chrono::seconds(60), abort_src);
+        }
+    });
+
+    if (first != tss.end()) {
+        co_return *std::prev(tss.end());
+    }
+
+    co_return std::nullopt;
+}
+
+future<> maybe_rewrite_streams_descriptions(
+        const database& db,
+        shared_ptr<db::system_distributed_keyspace> sys_dist_ks,
+        noncopyable_function<unsigned()> get_num_token_owners,
+        abort_source& abort_src) {
+    if (!db.has_schema(sys_dist_ks->NAME, sys_dist_ks->CDC_DESC_V1)) {
+        // This cluster never went through a Scylla version which used this table
+        // or the user deleted the table. Nothing to do.
+        co_return;
+    }
+
+    if (co_await db::system_keyspace::cdc_is_rewritten()) {
+        co_return;
+    }
+
+    // For each CDC log table get the TTL setting (from CDC options) and the table's creation time
+    std::vector<std::pair<db_clock::time_point, int>> times_and_ttls;
+    for (auto& [_, cf] : db.get_column_families()) {
+        auto& s = *cf->schema();
+        auto base = cdc::get_base_table(db, s.ks_name(), s.cf_name());
+        if (!base) {
+            // Not a CDC log table.
+            continue;
+        }
+        auto& cdc_opts = base->cdc_options();
+        if (!cdc_opts.enabled()) {
+            // This table is named like a CDC log table but it's not one.
+            continue;
+        }
+
+        times_and_ttls.emplace_back(as_timepoint(s.id()), cdc_opts.ttl());
+    }
```

Comment:
For a cluster with large number of tables this can cause reactor stalls.

Reply:
How many tables can a cluster really have? 1k? 10k? 100k?

Reply:
I think we had users with 10k tables before.
This should already be enough to cause problems.
We had problems related to the number of tables in the past.

Reply:
Processing a mutation fragment by `combined_reader` - which is a far more complex operation than what we're doing here - takes 100ns.
So processing 10k such fragments would take 100ns *10k = 1ms.
Now consider processing 10k tables by this loop, it'll take significantly less than 1ms.

I don't think this loop will cause reactor stalls.
Even if it did, they would be short, and this operation happens once per node, so I don't think it matters.

Reply:
ok.